migration/idempotent/999-last.sql

```diff
@@ -0,0 +1,271 @@
+        , 'prom_series'
+        , 'ps_tag'
+        , 'ps_trace'
+        )
+        ORDER BY nspname, proname
+    )
+    LOOP
+        EXECUTE format($$REVOKE ALL ON %s %I.%I(%s) FROM public$$, _rec.prokind, _rec.nspname, _rec.proname, _rec.args);
+        -- explicitly make the func/proc owned by the current_user (superuser). if (somehow) a malicious user predefined
+        -- a func/proc that we replace, we don't want them to be able to subsequently replace the body with malicious code
+        EXECUTE format($$ALTER %s %I.%I(%s) OWNER TO %I$$, _rec.prokind, _rec.nspname, _rec.proname, _rec.args, current_user);
+    END LOOP;
+END;
+$block$;
+
+-- make sure tables and sequences are pg_dump'ed properly
+DO $block$
+DECLARE
+    _sql text;
+BEGIN
+    FOR _sql IN
+    (
+        -- find tables and sequences which belong to the extension so we can mark them to have
+        -- their data dumped by pg_dump
+        SELECT format($$SELECT pg_catalog.pg_extension_config_dump('%I.%I', %L)$$,
+            n.nspname,
+            k.relname,
+            case (n.nspname, k.relname)
+                when ('_prom_catalog'::name, 'remote_commands'::name) then 'where seq >= 1000'
+                when ('_ps_trace'::name, 'tag_key'::name) then 'where id >= 1000'
+                else ''
+            end)
+        FROM pg_catalog.pg_depend d
+        INNER JOIN pg_catalog.pg_extension e ON (d.refobjid = e.oid)
+        INNER JOIN pg_catalog.pg_class k ON (d.objid = k.oid)
+        INNER JOIN pg_namespace n ON (k.relnamespace = n.oid)
+        WHERE d.refclassid = 'pg_catalog.pg_extension'::pg_catalog.regclass
+        AND d.deptype = 'e'
+        AND e.extname = 'promscale'
+        AND k.relkind IN ('r', 'p', 'S') -- tables AND sequences
+        AND (n.nspname, k.relname) NOT IN
+        (
+            -- these should NOT be config tables
+            -- migration table will be populated by the installation of extension
+            ('_ps_catalog'::name, 'migration'::name),
+            -- we want the telemetry data reset with restores
+            ('_ps_catalog'::name, 'promscale_instance_information'::name)
+        )
+        ORDER BY n.nspname, k.relname
+    )
+    LOOP
+        EXECUTE _sql;
+    END LOOP;
+END;
+$block$;
+
+-- make sure prom_admin can do a dump/restore
+DO $$
+DECLARE
+    _rec record;
+    _schema text;
+    _schemas text[] = ARRAY
+    [ '_prom_catalog'
+    , '_prom_ext'
+    , '_ps_catalog'
+    , '_ps_trace'
+    , 'prom_api'
+    , 'prom_data'
+    , 'prom_data_exemplar'
+    , 'prom_data_series'
+    , 'prom_info'
+    , 'prom_metric'
+    , 'prom_series'
+    , 'ps_tag'
+    , 'ps_trace'
+    ];
+BEGIN
+    -- grant schema privileges
+    FOREACH _schema IN ARRAY _schemas
+    LOOP
+        EXECUTE format('ALTER DEFAULT PRIVILEGES IN SCHEMA %I GRANT ALL PRIVILEGES ON TABLES TO prom_admin', _schema);
+        EXECUTE format('ALTER DEFAULT PRIVILEGES IN SCHEMA %I GRANT ALL PRIVILEGES ON SEQUENCES TO prom_admin', _schema);
```

Comment:
I am not exactly sure whether we need to do this as it will only apply to future tables and can't we just apply that when we create new tables? DEFAULT PRIVILEGES seem to be a weird pain in the ass as it's only applied when tables are created by certain users. I wonder if it's best to avoid and be explicit at table creation time.